File: rust/src/query.rs
use std::time;
use std::{
    collections::HashMap,
    io::{self, Write},
};

use crate::id_book::IDBookElement;
use crate::{file_skip_list, tokenizer::Tokenizer};
use std::fs::File;
use std::sync::{Arc, Mutex};
use std::thread;

pub const TOTAL_DOCUMENT_COUNT: u16 = 46843;

pub struct SearchEngine {
    query: String,
    tokens: Vec<String>,
    skiplists: Arc<Vec<Vec<file_skip_list::FileSkip>>>,
}

impl SearchEngine {
    pub fn new() -> Self {
        let mut skiplists = Vec::new();
        for i in 0..=9 {
            let skiplist = file_skip_list::FileSkip::read_skip_list((b'0' + i) as char);
            skiplists.push(skiplist);
        }
        for i in 0..26 {
            let skiplist = file_skip_list::FileSkip::read_skip_list((b'a' + i) as char);
            skiplists.push(skiplist);
        }
        Self {
            query: String::new(),
            tokens: Vec::new(),
            skiplists: Arc::new(skiplists),
        }
    }

    pub fn get_query(&mut self) {
        self.query.clear();
        print!("Enter your search query: ");
        io::stdout().flush().unwrap();
        io::stdin()
            .read_line(&mut self.query)
            .expect("Failed to read query");
        self.query = self.query.trim().to_string();
        self.tokens = Tokenizer::new().tokenize(&self.query);
    }

    pub fn set_query(&mut self, query: String) {
        self.query = query;
        self.tokens = Tokenizer::new().tokenize(&self.query);
    }

    pub fn search(&self) -> (Vec<String>, u128) {
        let time = time::Instant::now();
        println!("Searching for: \"{}\"", self.query);
        println!("Tokens: {:?}", self.tokens);

        // This will be shared across threads for adding candidates
        let candidates = Arc::new(Mutex::new(Vec::with_capacity(self.tokens.len())));
        let mut handles = vec![];

        for token in self.tokens.iter() {
            let candidates = Arc::clone(&candidates);
            let skiplists = Arc::clone(&self.skiplists);
            let token = token.clone();

            let handle = thread::spawn(move || {
                let first_char = token.chars().next().unwrap();
                let first_char_index = if first_char.is_ascii_digit() {
                    (first_char as u8 - b'0') as usize
                } else {
                    (first_char as u8 - b'a') as usize + 10
                };

                let offset_range =
                    file_skip_list::FileSkip::find_skip_entry(&skiplists[first_char_index], &token);

                let file_path = format!("inverted_index/merged/{}.txt", first_char);

                let mut candidate = Candidate::new(token.to_string());
                if let Ok(file) = File::open(&file_path) {
                    let postings =
                        file_skip_list::get_postings_from_offset_range(&file, offset_range, &token);
                    let posting_length = if token.len() <= 2 {
                        println!("Warning: Token '{}' is too short", token);
                        TOTAL_DOCUMENT_COUNT - 100
                    } else {
                        postings.postings.len() as u16
                    };
                    for single_posting in postings.postings {
                        let score = scoring_tf_idf(single_posting.term_freq, posting_length);
                        candidate.update_score(single_posting.doc_id, score);
                    }
                } else {
                    println!("Warning: Could not open index file for '{}'", first_char);
                }
                let mut candidates = candidates.lock().unwrap();
                candidates.push(candidate);
            });
            handles.push(handle);
        }

        for handle in handles {
            handle.join().unwrap();
        }

        if self.tokens.len() == 0 {
            return (Vec::new(), 0);
        }

        let mut candidates = Arc::try_unwrap(candidates).unwrap().into_inner().unwrap();
        if candidates.len() == 0 {
            return (Vec::new(), 0);
        }
        candidates.sort_by(|a: &Candidate, b: &Candidate| a.doc_ids.len().cmp(&b.doc_ids.len()));
        let mut base: HashMap<u16, f64> = candidates[0].doc_ids.clone();
        for i in 1..candidates.len() {
            let next = &candidates[i].doc_ids;
            base.retain(|k, _| next.contains_key(k));
        }

        let mut all_candidates: HashMap<u16, f64> = base;
        for candidate in candidates.iter() {
            for (doc_id, score) in candidate.doc_ids.iter() {
                all_candidates
                    .entry(*doc_id)
                    .and_modify(|s| *s += score)
                    .or_insert(*score);
            }
        }

        let mut sorted_candidates: Vec<(&u16, &f64)> = all_candidates.iter().collect();
        sorted_candidates.sort_by(|a, b| b.1.partial_cmp(a.1).unwrap());
        let mut results = Vec::new();
        let final_time = time.elapsed().as_millis();
        println!("Search took: {}ms", final_time);
        for (doc_id, score) in sorted_candidates.iter().take(10) {
            let doc = IDBookElement::get_doc_from_id(**doc_id);
            println!(
                "{}|> {}: {} (Score: {})",
                doc_id,
                doc.url,
                doc.path.display(),
                score
            );
            results.push(doc.url.clone());
        }
        (results, final_time)
    }
}

#[derive(Debug)]
pub struct Candidate {
    pub term: String,
    pub doc_ids: HashMap<u16, f64>, // for each doc_id, the tfidf score
}

impl Candidate {
    pub fn new(token: String) -> Self {
        Self {
            term: token,
            doc_ids: HashMap::new(),
        }
    }
    pub fn update_score(&mut self, doc_id: u16, score: f64) {
        self.doc_ids.insert(doc_id, score);
    }
}

pub fn scoring_tf_idf(term_freq: u16, posting_length: u16) -> f64 {
    let tf: f64 = f64::log2(term_freq as f64) + 1.0;
    let idf: f64 = f64::log2(TOTAL_DOCUMENT_COUNT as f64 / posting_length as f64);
    tf * idf
}

File: python/src/query.py
import time
import os
from collections import defaultdict
from typing import List
from tokenizer import Tokenizer
import math
import json

SRC_DIR = os.path.dirname(os.path.abspath(__file__))
OUTPUT_DIR = os.path.join(SRC_DIR, "merged_indexes")
PHONEBOOK = os.path.join(SRC_DIR, "phonebook.json")
TOTAL_DOCUMENT_COUNT = 40157  # Replace with the actual total document count

def get_postings(token):
    """
    Get the postings list for a given token from the inverted index.
    """
    first_char = token[0].lower()
    file_path = os.path.join(OUTPUT_DIR, f"{first_char}.txt")
    postings = {}
    if os.path.exists(file_path):
        with open(file_path, 'r', encoding='utf-8') as f:
            for line in f:
                parts = line.strip().split()
                if parts[0] == token:
                    for posting in parts[1:]:
                        doc_id, freq = map(
                            int, posting.removesuffix(',').split(':'))
                        postings[doc_id] = freq
                    break  # stop after finding the token
    return postings

def scoring_tf_idf(term_freq: int, posting_length: int) -> float:
    """
    Calculate the TF-IDF score for a term in a document.
    """
    tf = (math.log2(term_freq) + 1.0 )
    idf = math.log2(TOTAL_DOCUMENT_COUNT / posting_length)
    return tf * idf

class Candidate:
    """
    A candidate document for a query.
    """

    def __init__(self, doc_id):
        self.doc_id = doc_id
        self.tokens_matched = {}   
        self.total_score = 0.0   
        self.tf_idf = 0.0 
        self.info = {}  # frequency, posting_length, token
    
    def set_info(self, frequency, posting_length, token):
        self.info[token] = (frequency, posting_length)
    
    def update_score(self, token, score):
        self.tokens_matched[token] = score
        self.total_score += score

    def has_all_tokens(self, query_tokens):
        return all(token in self.tokens_matched for token in query_tokens)
    
    def get_total_score(self):
        return self.total_score
        
    def get_total_tf_idf(self):
        return self.tf_idf
    
    # def compute_tf_idf(self):
    #     for token in self.info.keys():
    #         self.compute_tf_idf_for_token(token)
    #         term_freq, posting_length = self.info[token]
    #         total_tokens_in_a_doc = get_num_tokens_of_a_doc(self.doc_id)
    #         tf = (math.log2(term_freq/total_tokens_in_a_doc) + 1.0)
    #         idf = math.log2(TOTAL_DOCUMENT_COUNT / posting_length)
    #         self.tf_idf += tf * idf
    #     return self.tf_idf        

    


class SearchEngine:
    """
    A search engine that takes a query and returns the matching documents.
    """

    def __init__(self):
        self.query = ""
        self.tokens = []
        self.time = 0

    def get_query(self):
        """
        Get the query from the user through the console.
        """
        self.query = input("Enter your query: ").strip()
        tokenizer = Tokenizer()
        self.tokens = tokenizer.tokenize(self.query)

    def set_query(self, query):
        """
        Set the query from the user.
        """
        self.query = query
        tokenizer = Tokenizer()
        self.tokens = tokenizer.tokenize(self.query)

    def search(self) -> List[str]:
        start_time = time.time()
        print(f"Searching...: {self.query}")
        print(f"Tokens: {self.tokens}")
        candidates = {}

        for token in self.tokens:
            # get the postings list for the token
            postings = get_postings(token)
            postings_len = len(postings)
            for cur_doc, frequency in postings.items():
                doc_id = cur_doc + 1
                if doc_id not in candidates:
                    candidates[doc_id] = Candidate(doc_id)
                
                score = scoring_tf_idf(frequency, postings_len)
                candidates[doc_id].update_score(token, score)
                # candidates[doc_id].set_info(frequency, postings_len, token)
        # Filter candidates to only those that match all query tokens
        valid_candidates = get_valid_candidates(candidates, self.tokens)
        finished_time = int((time.time() - start_time) * 1000)
        self.time = finished_time          
        valid_candidates.sort(key=lambda c: c.get_total_score(), reverse=True)
        # valid_candidates.sort(key=lambda c: c.get_tf_idf(), reverse=True)

        # Take only top 5 results if available
        valid_candidates = valid_candidates[:5]
        doc_ids = []
        if valid_candidates:
            print("Matching Documents:")
            for i, candidate in enumerate(valid_candidates):
                print(
                    f"{i+1}. Document ID: {candidate.doc_id}, Score: {candidate.tokens_matched[token]}")
                doc_ids.append(candidate.doc_id)
        else:
            return [] 
        print(f"Search completed in {finished_time} ms.")
        
        return get_doc_info(doc_ids)
        
    def get_time(self):
        return self.time

def get_valid_candidates(candidates, query_tokens):
    valid_candidates = []
    for c in candidates.values():
        if c.has_all_tokens(query_tokens):
            valid_candidates.append(c)
    return valid_candidates

def get_doc_info(doc_ids):
    """
    Get the document information for a given document ID.
    Returns an array of a docID's URL and content
    First element has the highest score. 
    """
    with open(PHONEBOOK, 'r') as f:
        phonebook = json.load(f)
    
    result = []
    
    for doc_id in doc_ids:
        file_path = phonebook[str(doc_id)]
        print(f"File path: {file_path}")
        with open(file_path, 'r') as doc_file:
            doc_data = json.load(doc_file)
            result.append((doc_data["url"], doc_data["content"]))
    
    return result

# def get_num_tokens_of_a_doc(doc_id):
#     """
#     Get the document information for a given document ID.
#     Returns an array of a docID's URL and content
#     First element has the highest score. 
#     """
#     with open(PHONEBOOK, 'r') as f:
#         phonebook = json.load(f)
    
#     tokenizer = Tokenizer()    
#     file_path = phonebook[str(doc_id)]
#     print(f"File path: {file_path}")
#     with open(file_path, 'r') as doc_file:
#         doc_data = json.load(doc_file)
#         return len(tokenizer.tokenize(doc_data["content"]))
                
